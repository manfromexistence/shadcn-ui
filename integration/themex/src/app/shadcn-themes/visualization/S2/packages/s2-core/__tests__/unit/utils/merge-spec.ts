import { LayoutWidthType, type S2DataConfig } from '@/common';
import { customMerge, setupDataConfig, setupOptions } from '@/utils/merge';

describe('merge test', () => {
  test('should replace old array with new one', () => {
    expect(customMerge({ arr: [1, 2, 3] }, { arr: [4, 5, 6] })).toEqual({
      arr: [4, 5, 6],
    });
  });

  test('should force replace empty array', () => {
    expect(customMerge({ arr: [1, 2, 3] }, { arr: [] })).toEqual({
      arr: [],
    });
  });

  test('should return new object', () => {
    const obj = { name: 'name' };
    const result = customMerge(obj, { age: 100 });

    expect(obj).toEqual({ name: 'name' });
    expect(result).toEqual({ name: 'name', age: 100 });
  });

  test('should get safety data config', () => {
    expect(setupDataConfig(null)).toMatchSnapshot();
  });

  test('should unique dataConfig fields', () => {
    expect(
      setupDataConfig({
        fields: {
          rows: ['province', 'city', 'city'],
          columns: ['type', 'type'],
          values: ['count', 'cost', 'cost'],
          valueInCols: false,
        },
      }),
    ).toMatchSnapshot();
  });

  test('should cancel valueInCols if custom rows is not empty by get safety data config', () => {
    const rows = [{ field: '1', title: 'test' }];
    const fields: Partial<S2DataConfig['fields']> = {
      rows,
      values: ['1'],
      valueInCols: true,
    };

    expect(
      setupDataConfig({
        fields,
      }),
    ).toMatchSnapshot();
  });

  test('should cancel valueInCols if value is empty by get safety data config', () => {
    const fields: Partial<S2DataConfig['fields']> = {
      values: [],
      valueInCols: true,
    };

    expect(
      setupDataConfig({
        fields,
      }),
    ).toMatchSnapshot();
  });

  test('should merge old dataCfg', () => {
    const oldDataCfg: Partial<S2DataConfig> = {
      data: [{ value: 1 }, { value: 2 }],
    };
    const fields: Partial<S2DataConfig['fields']> = {
      values: [],
      valueInCols: true,
    };

    expect(
      setupDataConfig(oldDataCfg, {
        fields,
      }),
    ).toMatchSnapshot();
  });

  test('should get safety options', () => {
    // 加这个测试可以防止 本地跑 demo 修改了默认配置 直接提交
    expect(setupOptions(null)).toMatchSnapshot();
  });

  test('should not setup correctly compact layout width type style', () => {
    expect(
      setupOptions({
        style: {
          layoutWidthType: LayoutWidthType.Compact,
        },
      }).style,
    ).toMatchSnapshot();
  });

  test('should get custom options', () => {
    const options = setupOptions({
      tooltip: {
        enable: false,
        operation: {
          sort: false,
          menu: {
            items: [
              {
                key: 'custom',
                label: 'custom',
              },
            ],
          },
        },
      },
    });

    expect(options.tooltip).toMatchSnapshot();
  });

  test('should get custom data config', () => {
    const dataConfig = setupDataConfig({
      fields: {
        rows: ['test'],
        values: ['value'],
      },
    });

    expect(dataConfig.fields).toStrictEqual({
      rows: ['test'],
      columns: [],
      values: ['value'],
      valueInCols: true,
    });
  });
});
