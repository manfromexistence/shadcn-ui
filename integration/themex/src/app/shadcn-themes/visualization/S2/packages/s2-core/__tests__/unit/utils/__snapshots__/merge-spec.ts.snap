// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge test should cancel valueInCols if custom rows is not empty by get safety data config 1`] = `
Object {
  "data": Array [],
  "fields": Object {
    "columns": Array [],
    "rows": Array [
      Object {
        "field": "1",
        "title": "test",
      },
    ],
    "valueInCols": false,
    "values": Array [
      "1",
    ],
  },
  "filterParams": Array [],
  "meta": Array [],
  "sortParams": Array [],
}
`;

exports[`merge test should cancel valueInCols if value is empty by get safety data config 1`] = `
Object {
  "data": Array [],
  "fields": Object {
    "columns": Array [],
    "rows": Array [],
    "valueInCols": false,
    "values": Array [],
  },
  "filterParams": Array [],
  "meta": Array [],
  "sortParams": Array [],
}
`;

exports[`merge test should get custom options 1`] = `
Object {
  "autoAdjustBoundary": "body",
  "enable": false,
  "operation": Object {
    "hiddenColumns": false,
    "menu": Object {
      "items": Array [
        Object {
          "key": "custom",
          "label": "custom",
        },
      ],
    },
    "sort": false,
  },
}
`;

exports[`merge test should get safety data config 1`] = `
Object {
  "data": Array [],
  "fields": Object {
    "columns": Array [],
    "rows": Array [],
    "valueInCols": false,
    "values": Array [],
  },
  "filterParams": Array [],
  "meta": Array [],
  "sortParams": Array [],
}
`;

exports[`merge test should get safety options 1`] = `
Object {
  "conditions": Object {},
  "cornerExtraFieldText": "",
  "cornerText": "",
  "customSVGIcons": Array [],
  "debug": false,
  "device": "pc",
  "frozen": Object {
    "colCount": 0,
    "rowCount": 0,
    "rowHeader": true,
    "trailingColCount": 0,
    "trailingRowCount": 0,
  },
  "hd": true,
  "headerActionIcons": Array [],
  "height": 480,
  "hierarchyType": "grid",
  "interaction": Object {
    "autoResetSheetStyle": true,
    "brushSelection": Object {
      "colCell": true,
      "dataCell": true,
      "rowCell": true,
    },
    "copy": Object {
      "enable": true,
      "withFormat": true,
      "withHeader": false,
    },
    "eventListenerOptions": false,
    "hiddenColumnFields": Array [],
    "hoverFocus": Object {
      "duration": 800,
    },
    "hoverHighlight": true,
    "linkFields": Array [],
    "multiSelection": true,
    "overscrollBehavior": "auto",
    "rangeSelection": true,
    "resize": Object {
      "colCellHorizontal": true,
      "colCellVertical": true,
      "colResizeType": "current",
      "cornerCellHorizontal": true,
      "minCellHeight": 20,
      "minCellWidth": 40,
      "rowCellVertical": true,
      "rowResizeType": "current",
    },
    "scrollSpeedRatio": Object {
      "horizontal": 1,
      "vertical": 1,
    },
    "scrollbarPosition": "content",
    "selectedCellHighlight": false,
    "selectedCellsSpotlight": false,
  },
  "placeholder": Object {
    "cell": "-",
    "empty": Object {
      "icon": "Empty",
    },
  },
  "seriesNumber": Object {
    "enable": false,
  },
  "showDefaultHeaderActionIcon": false,
  "style": Object {
    "colCell": Object {
      "height": 30,
      "maxLines": 1,
      "textOverflow": "ellipsis",
      "wordWrap": true,
    },
    "cornerCell": Object {
      "maxLines": 1,
      "textOverflow": "ellipsis",
      "wordWrap": true,
    },
    "dataCell": Object {
      "height": 30,
      "maxLines": 1,
      "textOverflow": "ellipsis",
      "width": 96,
      "wordWrap": true,
    },
    "layoutWidthType": "adaptive",
    "rowCell": Object {
      "maxLines": 1,
      "showTreeLeafNodeAlignDot": false,
      "textOverflow": "ellipsis",
      "wordWrap": true,
    },
    "seriesNumberCell": Object {
      "maxLines": 1,
      "textOverflow": "ellipsis",
      "wordWrap": true,
    },
  },
  "tooltip": Object {
    "autoAdjustBoundary": "body",
    "enable": false,
    "operation": Object {
      "hiddenColumns": false,
      "menu": Object {
        "items": Array [],
      },
      "sort": false,
    },
  },
  "totals": Object {},
  "width": 600,
}
`;

exports[`merge test should merge old dataCfg 1`] = `
Object {
  "data": Array [
    Object {
      "value": 1,
    },
    Object {
      "value": 2,
    },
  ],
  "fields": Object {
    "columns": Array [],
    "rows": Array [],
    "valueInCols": false,
    "values": Array [],
  },
  "filterParams": Array [],
  "meta": Array [],
  "sortParams": Array [],
}
`;

exports[`merge test should not setup correctly compact layout width type style 1`] = `
Object {
  "colCell": Object {
    "height": 30,
    "maxLines": 1,
    "textOverflow": "ellipsis",
    "wordWrap": true,
  },
  "cornerCell": Object {
    "maxLines": 1,
    "textOverflow": "ellipsis",
    "wordWrap": true,
  },
  "dataCell": Object {
    "height": 30,
    "maxLines": 1,
    "textOverflow": "ellipsis",
    "width": 96,
    "wordWrap": true,
  },
  "layoutWidthType": "compact",
  "rowCell": Object {
    "maxLines": 1,
    "showTreeLeafNodeAlignDot": false,
    "textOverflow": "ellipsis",
    "wordWrap": true,
  },
  "seriesNumberCell": Object {
    "maxLines": 1,
    "textOverflow": "ellipsis",
    "wordWrap": true,
  },
}
`;

exports[`merge test should unique dataConfig fields 1`] = `
Object {
  "data": Array [],
  "fields": Object {
    "columns": Array [
      "type",
    ],
    "rows": Array [
      "province",
      "city",
    ],
    "valueInCols": false,
    "values": Array [
      "count",
      "cost",
    ],
  },
  "filterParams": Array [],
  "meta": Array [],
  "sortParams": Array [],
}
`;
